### PR TITLE
Use gunicorn

### PR DESCRIPTION
**Pull request by foriequal0**
Refactored the code to ease the adoption of the gunicorn, and introduced the usage of the gunicorn in Readme.

This is not a destructive change, and the gunicorn is optional.

### CODE REVIEW

File: proxy.py [hunk 1]
```diff
@@ -24,23 +24,19 @@
     args = parser.parse_args()
     app = create(**vars(args))
     app.run(host=args.bind, port=args.port, debug=args.debug)
-
+
```
Comment (sgkim126) [inline proxy.py]: redundant spaces

**Review by sgkim126 — CHANGES_REQUESTED** [review]

**Comment by sgkim126** [comment]
Why do you use gunicorn?

**Comment by sgkim126** [comment]
Please merge 9a6bb67 to cb0901883a7b2fb97e52a84910741396d9082f44

**Comment by foriequal0** [comment]
You can use other WSGI containers: http://flask.pocoo.org/docs/1.0/deploying/wsgi-standalone/
The gunicorn is the only option that is the almost pure python and I can easily find an example that launches a server without modifying existing code.

**Comment by foriequal0** [comment]
ping.lua
```
wrk.method = "POST"
wrk.body = "{\"jsonrpc\": \"2.0\", \"method\": \"ping\", \"params\": [], \"id\": null}"
wrk.headers["Content-Type"] = "application/json"
```

```
wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8080 # codechain
Running 30s test @ http://192.168.100.6:8080
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   115.70ms  147.26ms   1.88s    94.49%
    Req/Sec   351.48    126.61   660.00     74.80%
  117596 requests in 30.10s, 19.18MB read
  Socket errors: connect 0, read 0, write 0, timeout 407
Requests/sec:   3906.56
Transfer/sec:    652.37KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8081 # python proxy.py
Running 30s test @ http://192.168.100.6:8081
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   862.27ms  330.51ms   1.99s    87.56%
    Req/Sec    19.67     12.02    80.00     82.31%
  6605 requests in 30.10s, 1.25MB read
  Socket errors: connect 0, read 6, write 0, timeout 538
Requests/sec:    219.43
Transfer/sec:     42.43KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8082 # gunicorn 
Running 30s test @ http://192.168.100.6:8082
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   550.21ms  302.17ms   1.98s    84.90%
    Req/Sec    28.45     18.28   130.00     57.17%
  5924 requests in 30.08s, 1.15MB read
  Socket errors: connect 0, read 30, write 0, timeout 539
Requests/sec:    196.95
Transfer/sec:     39.24KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8083 # gunicorn -k eventlet
Running 30s test @ http://192.168.100.6:8083
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.10s   454.78ms   2.00s    63.76%
    Req/Sec    28.88     15.77   168.00     67.62%
  9968 requests in 30.10s, 1.99MB read
  Socket errors: connect 0, read 0, write 0, timeout 557
Requests/sec:    331.18
Transfer/sec:     67.60KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8084 # gunicorn -k gevent
Running 30s test @ http://192.168.100.6:8084
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.34s   253.24ms   2.00s    72.63%
    Req/Sec    24.19     18.11   125.00     76.76%
  7456 requests in 30.10s, 1.49MB read
  Socket errors: connect 0, read 0, write 0, timeout 1391
Requests/sec:    247.74
Transfer/sec:     50.56KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8085 # gunicorn -w4
Running 30s test @ http://192.168.100.6:8085
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   457.68ms  477.40ms   2.00s    82.90%
    Req/Sec    49.60     39.76   240.00     66.19%
  10718 requests in 30.07s, 2.09MB read
  Socket errors: connect 0, read 51, write 0, timeout 542
Requests/sec:    356.39
Transfer/sec:     71.00KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8085 # gunicorn -w4 -k eventlet
Running 30s test @ http://192.168.100.6:8085
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   486.44ms  501.65ms   2.00s    82.48%
    Req/Sec    43.75     34.23   190.00     61.85%
  9256 requests in 30.08s, 1.80MB read
  Socket errors: connect 0, read 16, write 0, timeout 578
Requests/sec:    307.68
Transfer/sec:     61.30KB

wrk -t12 -c400 -d30s -s ping.lua http://192.168.100.6:8085 # gunicorn -w4 -k gevent
Running 30s test @ http://192.168.100.6:8085
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   384.10ms  423.66ms   2.00s    88.38%
    Req/Sec    54.34     39.39   250.00     66.25%
  12868 requests in 30.09s, 2.50MB read
  Socket errors: connect 0, read 13, write 0, timeout 823
Requests/sec:    427.68
Transfer/sec:     85.20KB
```
🤔

**Comment by foriequal0** [comment]
https://github.com/foriequal0/codechain-proxy/blob/feature/aiohttp/proxy.py
aiohttp version is better, but still slow.
```
% python proxy.py --port 8080 --forward 8081 2>/dev/null
```
```
[0] % wrk -t12 -c400 -d30s -s ping.lua http://localhost:8080
Running 30s test @ http://localhost:8080
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   637.13ms  146.64ms   1.83s    87.27%
    Req/Sec    51.60     33.21   191.00     72.80%
  16448 requests in 30.08s, 3.18MB read
  Socket errors: connect 0, read 261, write 0, timeout 0
  Non-2xx or 3xx responses: 500
Requests/sec:    546.88
Transfer/sec:    108.37KB
```

**Comment by sgkim126** [comment]
Thank you @foriequal0 
IMO, it's better to use another implementation instead of Python.

**Pull request closed without merging**